_posts/2016-04-06-intro-to-macros.md

```diff
@@ -0,0 +1,111 @@
+
+# Metaprogramming in Elixir
+
+_Note:_ If you are already familiar and experienced with Elixir Metaprogramming, you can jump to [last section](#last-section) as it lists Metaprogramming resources
+and constructs used in upcoming blogs when creating different extension DSL's.
+
+**Elixir _docs_** are excellent resources to get familiar with Metaprogramming:
+
+- [Quote and unquote](http://elixir-lang.org/getting-started/meta/quote-and-unquote.html)
+- [Macros](http://elixir-lang.org/getting-started/meta/macros.html)
+- [Domain Specific Languages](http://elixir-lang.org/getting-started/meta/domain-specific-languages.html)
+
+Check out **Understanding Elixir Macros _blog_ series** by [@sasajuric](https://twitter.com/sasajuric)
+
+- [Understanding Elixir Macros, Part 1 - Basics](http://theerlangelist.com/article/macros_1)
+- [Understanding Elixir Macros, Part 2 - Micro Theory](http://theerlangelist.com/article/macros_2)
+- [Understanding Elixir Macros, Part 3 - Getting into the AST](http://theerlangelist.com/article/macros_3)
+- [Understanding Elixir Macros, Part 4 - Diving Deeper](http://theerlangelist.com/article/macros_4)
+- [Understanding Elixir Macros, Part 5 - Reshaping the AST](http://theerlangelist.com/article/macros_5)
+- [Understanding Elixir Macros, Part 6 - In-place Code Generation](http://theerlangelist.com/article/macros_6)
+
+[**_Book_** Metaprogramming Elixir](https://pragprog.com/book/cmelixir/metaprogramming-elixir) by [@chris_mccord](https://twitter.com/chris_mccord) is next step for deep dive into metaprogramming.
+
+>
+Why another tutorial on already well-documented metaprogramming topic ?
+>
+**To revise and refresh something that we would refer time and again when reviewing Model, Router, View extension DSLs**
+
+Extension DSLs will be using below constructs to get the job done :)
+
+- [Module.register_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#register_attribute/3)
+  - Registers an attribute. By registering an [attribute](http://elixir-lang.org/getting-started/module-attributes.html), a developer is able to customize how Elixir will store and accumulate the attribute values.
+- [Module.put_attribute/3](http://elixir-lang.org/docs/stable/elixir/Module.html#put_attribute/3)
+  - Puts an Erlang attribute to the given module with the given key and value
+- [@before_compile](http://elixir-lang.org/docs/stable/elixir/Module.html)
+  - A hook `__before_compile__/1` that will be invoked before the module is compiled
+  - It allows us to inject code into the module when its definition is complete
+- [\_\_using\_\_ hook](http://elixir-lang.org/docs/stable/elixir/Kernel.html#use/2)
+  - Find the documentation under the `examples for the usage, context and best practices` in link above
```

Comment:
Find the documentation under title `examples for the usage, context and best practices` in link above